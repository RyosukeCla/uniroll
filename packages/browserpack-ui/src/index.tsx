export { EnvContext, useEnv } from "./components/contexts";
export { App } from "./components/App";
