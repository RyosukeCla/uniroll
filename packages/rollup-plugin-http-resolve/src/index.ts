import { Plugin } from "rollup";
import path from "path";

export type ImportMaps = {
  imports: { [k: string]: string };
};

function isHttpProtocol(id: string) {
  return id.startsWith("http://") || id.startsWith("https://");
}

const DEBUG = false;
const log = (...args: any) => DEBUG && console.log(...args);

type HttpResolveOptions = {
  cache?: any;
  fetcher?: (url: string) => Promise<string>;
  onRequest?: (url: string) => void;
  onUseCache?: (url: string) => void;
  transform?: (code: string) => string;
  fallback?: (
    id: string,
    importer: string,
    onWarn: (warn: any) => void
  ) => Promise<string | void> | void | string;
};
const defaultCache = new Map();
export const httpResolve = function httpResolve_({
  cache = defaultCache,
  onRequest,
  transform,
  onUseCache,
  fetcher,
  fallback,
}: HttpResolveOptions) {
  return {
    name: "http-resolve",
    async resolveId(id: string, importer: string) {
      log("[http-resolve:resolveId:enter]", id, "from", importer);
      // on network resolve
      if (importer && isHttpProtocol(importer)) {
        if (id.startsWith("https://")) {
          log("[http-reslove:end] return with https", id);
          return id;
        }
        const { pathname, protocol, host } = new URL(importer);
        if (id.startsWith("/")) {
          // /_/...
          log(
            "[http-reslove:end] return with host root",
            `${protocol}//${host}${id}`
          );
          return `${protocol}//${host}${id}`;
        } else if (id.startsWith(".")) {
          // ./xxx/yyy
          // relative path
          const resolvedPathname = path.join(path.dirname(pathname), id);
          const newId = `${protocol}//${host}${resolvedPathname}`;
          log("[http-resolve:end] return with relativePath", newId);
          return newId;
        }
      }
      if (fallback) {
        const fallbacked = await fallback(id, importer, this.warn);
        log("[http-resolve:end] use fallback to", id, "=>", fallbacked);
        if (fallbacked) {
          return fallbacked;
        }
      }
    },
    async load(id: string) {
      log("[http-resolve:load]", id);
      if (id === null) {
        this.warn("irregular missing id");
        return;
      }
      if (isHttpProtocol(id)) {
        const cached = await cache.get(id);
        if (cached) {
          onUseCache?.(id);
          return cached;
        }
        onRequest?.(id);
        if (fetcher) {
          let code = await fetcher(id);
          code = transform?.(code) ?? code;
          await cache.set(id, code);
          return code;
        } else {
          const res = await fetch(id);
          if (!res.ok) {
            throw res.statusText;
          }
          let code = await res.text();
          code = transform?.(code) ?? code;
          await cache.set(id, code);
          return code;
        }
      }
    },
  } as Plugin;
};

export function createImportMapsFallback({
  importmaps,
}: {
  importmaps?: ImportMaps;
}) {
  return (
    id: string,
    importer: string | void = undefined,
    warn: (warning: any) => any
  ): Promise<void | string> | void | string => {
    if (importer == null) {
      return;
    }
    if (id.startsWith("http")) {
      return;
    }
    if (id.startsWith(".")) {
      return;
    }

    const mapped = importmaps?.imports[id];
    if (mapped) {
      return mapped;
    }
    warn(`missed fallback to https://cdn.skypack.dev/${id}`);
    return `https://cdn.skypack.dev/${id}`;
  };
}

// export default httpResolve;
