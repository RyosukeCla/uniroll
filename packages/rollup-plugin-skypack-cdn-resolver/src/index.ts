import { Plugin } from "rollup";
import path from "path";
const SKYPACK_CDN_HOST = "https://cdn.skypack.dev";

export function skypackCDNResolver({
  cache = new Map(),
  onRequest,
  onUseCache,
  ignorePolyfill = false,
}: {
  cache?: any;
  ignorePolyfill?: boolean;
  onRequest?: (url: string) => void;
  onUseCache?: (url: string) => void;
}) {
  return {
    async resolveId(id: string, importer: string) {
      // console.log(id, importer);
      if (importer && importer.startsWith(SKYPACK_CDN_HOST)) {
        // load Skypack in Skypack
        if (id.startsWith(SKYPACK_CDN_HOST)) {
          return id;
        }

        const { pathname, protocol, host } = new URL(importer);
        if (id.startsWith("/")) {
          return `${protocol}//${host}${id}`;
        } else {
          const resolvedPathname = path.join(path.dirname(pathname), id);
          const newId = `${protocol}//${host}${resolvedPathname}`;
          return newId;
        }
      }
    },
    async load(id: string) {
      if (id.includes("@pika/polyfill")) {
        return `// ignored: ${id}`;
      }
      if (id.startsWith(SKYPACK_CDN_HOST)) {
        const cached = await cache.get(id);
        if (cached) {
          onUseCache && onUseCache(id);
          return cached;
        }
        onRequest && onRequest(id);
        const res = await fetch(id);
        if (!res.ok) {
          throw res.statusText;
        }
        const code = await res.text();
        await cache.set(id, code);
        return code;
      }
    },
  } as Plugin;
}
