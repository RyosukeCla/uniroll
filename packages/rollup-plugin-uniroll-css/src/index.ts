import { Options } from "./types";
import { Plugin } from "rollup";
import { transformCss } from "./transformCss";
import { transformWithAutoprefixer } from "./transformWithAutoprefixer";
import { wrapWithStyleInjector } from "./wrapWithInjector";

export const css = (options: Options = {}) => {
  return {
    async transform(code: string, id: string): Promise<string | void> {
      if (id.endsWith(".css")) {
        const css = await transformCss(code, options);
        const autoprefixedCss = await transformWithAutoprefixer(css);
        return wrapWithStyleInjector(autoprefixedCss);
      }
    }
  } as Plugin;
};
