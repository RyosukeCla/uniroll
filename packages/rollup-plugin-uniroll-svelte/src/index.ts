import { PreprocessorGroup } from "svelte/types/compiler/preprocess";
import type { Plugin } from "rollup";
import { compile as svelteCompile, preprocess } from "svelte/compiler";
import {
  cdnRewriteTransformerFactory,
  createSveltePreprocessor,
} from "./tsPreprocess";
import ts from "typescript";
import { CompileOptions } from "svelte/types/compiler/interfaces";
import { ResolveIdFallback } from "./types";

type SveltePluginOptions = {
  target: ts.ScriptTarget;
  resolveIdFallback?: ResolveIdFallback;
  svelteOptions?: CompileOptions;
  extraPreprocess?: PreprocessorGroup[];
};

const defaultResolveIdFallback: ResolveIdFallback = (
  id: string,
  importer?: string
) => {
  return;
};

export const svelte = (opts: SveltePluginOptions) => {
  /**
   * if css is set to false, then output css file instead of js injection.
   */
  const willOutptuCss = !opts.svelteOptions?.css;
  const cssFiles = new Map<string, any>();

  return {
    name: "uniroll-svelte",
    buildStart() {
      cssFiles.clear();
    },
    load(id: string) {
      if (!cssFiles.has(id)) return;
      return cssFiles.get(id);
    },
    async transform(code: string, id: string) {
      if (id.endsWith(".svelte")) {
        const tsPreprocess = createSveltePreprocessor({
          target: opts.target,
          importer: id,
          resolveIdFallback: opts.resolveIdFallback ?? defaultResolveIdFallback,
        });

        const { code: preprocessed } = await preprocess(
          code,
          [tsPreprocess, ...(opts.extraPreprocess ?? ([] as any))],
          {
            filename: id,
          }
        );
        const result = svelteCompile(preprocessed, opts.svelteOptions);
        if (result.warnings.length > 0) {
          this.warn(result.warnings.map((t) => t.message).join("\n"));
        }

        if (willOutptuCss && result.css.code) {
          const cssFileName = `/svelte-generated-css/${Math.random().toString(16)}.css`;
          result.js.code += `\nimport "${cssFileName}";\n`;
          cssFiles.set(cssFileName, result.css);
        }

        const ret = ts.transpileModule(result.js.code, {
          fileName: "$temp.tsx",
          compilerOptions: {
            target: opts.target,
            module: ts.ModuleKind.ESNext,
          },
          transformers: {
            before: [cdnRewriteTransformerFactory(opts.resolveIdFallback!, id)],
          },
        });

        return {
          code: ret.outputText,
          map: ret.sourceMapText,
        };
      }
    },
  } as Plugin;
};
