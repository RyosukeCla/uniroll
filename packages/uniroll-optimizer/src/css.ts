import cssnano from "cssnano";
// @ts-ignore
import precss from "precss";
import postcss from "postcss";
import autoprefixer from "autoprefixer";

export function optimize(css: string) {
  return transformWithAutoprefixer(css);
}

async function transformWithAutoprefixer(input: string) {
  const result = await postcss([
    precss,
    autoprefixer({
      // @ts-ignore
      grid: true,
      overrideBrowserslist: ["last 2 versions", "ie 11"]
    }),
    cssnano({
      // @ts-ignore
      autoprefixer: false
    })
  ]).process("/* autoprefixer grid: autoplace */\n" + input, {
    from: undefined
  });
  return result.css;
}

async function main() {
  try {
    const ret = await optimize(`body { color: red;  }`);
    console.log(ret);
  } catch (err) {
    console.error(err);
  }
}

main();
