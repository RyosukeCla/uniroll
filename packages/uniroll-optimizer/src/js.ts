import terser from "terser";

export function optimize(js: string) {
  return terser.minify(js, { module: true }).code;
}

function main() {
  const code = optimize(`const a = 1 + 1;\nexport default a`);
  console.log(code);
}

main();
