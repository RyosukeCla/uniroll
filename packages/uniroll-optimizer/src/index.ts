import { optimize as optimizeCss } from "./css";
import { optimize as optimizeJs } from "./js";
export async function optimizeChunks(files: {
  [k: string]: string;
}): Promise<{ [k: string]: string }> {
  return Object.fromEntries(
    await Promise.all(
      Object.entries(files).map(async ([fpath, value]) => {
        if (fpath.endsWith(".js")) {
          return [fpath, await optimizeJs(value)];
        }
        if (fpath.endsWith(".css")) {
          return [fpath, await optimizeCss(value)];
        }
        return [fpath, value];
      })
    )
  );
}
export { optimizeCss, optimizeJs };
