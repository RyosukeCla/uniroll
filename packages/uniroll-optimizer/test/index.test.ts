import { optimizeChunks } from "../src/index";
// How to use
// function main() {
//   const code = optimize(`const a = 1 + 1;\nexport default a`);
//   console.log(code);
// }

// main();

test("optimize", async () => {
  const ret = await optimizeChunks({
    "index.js": "console.log(true);",
    "sytle.css": "div { box-sizing: border-box; }",
  });

  expect(ret).toMatchSnapshot();
});
