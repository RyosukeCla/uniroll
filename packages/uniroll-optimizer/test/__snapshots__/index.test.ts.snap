// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`optimize 1`] = `
Object {
  "index.js": "console.log(!0);",
  "sytle.css": "div{-webkit-box-sizing:border-box;box-sizing:border-box}",
}
`;
